### PR TITLE
MINIFICPP-1307 - Add attribute-based merge test

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/test/archive-tests/MergeFileTests.cpp [hunk 1]
```diff
@@ -752,3 +752,80 @@ TEST_CASE("MergeFileZip", "[mergefiletest5]") {
...
+    session->commit();
+  }
+  // validate the merge content
+  std::set<std::shared_ptr<core::FlowFile>> expiredFlowRecords;
+  std::shared_ptr<core::FlowFile> flow1 = output->poll(expiredFlowRecords);
+  std::shared_ptr<core::FlowFile> flow2 = output->poll(expiredFlowRecords);
+  {
+    FixedBuffer callback(flow1->getSize());
+    sessionGenFlowFile.read(flow1, &callback);
+    std::ifstream file1(EXPECT_MERGE_CONTENT_FIRST, std::ios::binary);
```
Comment (fgerlits) [inline libminifi/test/archive-tests/MergeFileTests.cpp:819]: What is the point of writing these `EXPECT_MERGE_CONTENT_*` files to disk and then reading them back?  We could use strings `EXPECT_MERGE_CONTENT_FIRST = "024"`, `EXPECT_MERGE_CONTENT_SECOND = "135"`, and compare that to `callback.to_string()`.

Comment (adamdebreceni) [reply libminifi/test/archive-tests/MergeFileTests.cpp:819]: agree that it is strange, for now consistency wins, we should definitely refactor it, as the CompressContentTests do the same

Comment (adamdebreceni) [reply libminifi/test/archive-tests/MergeFileTests.cpp:819]: created a [ticker](https://issues.apache.org/jira/browse/MINIFICPP-1308)

File: libminifi/test/archive-tests/MergeFileTests.cpp [hunk 2]
```diff
@@ -752,3 +752,80 @@ TEST_CASE("MergeFileZip", "[mergefiletest5]") {
...
+    else
+      flow->setAttribute("tag", "odd");
+    record[i] = flow;
+  }
+  input->put(record[0]);
+  input->put(record[1]);
+  input->put(record[2]);
+  input->put(record[3]);
+  input->put(record[4]);
+  input->put(record[5]);
```
Comment (hunyadi-dev) [inline libminifi/test/archive-tests/MergeFileTests.cpp:802]: Why can't you do this in the loop above?

Comment (adamdebreceni) [reply libminifi/test/archive-tests/MergeFileTests.cpp:802]: for consistency, all tests in this file separate the creation of flowFiles from their placement into the respective connection

File: libminifi/test/archive-tests/MergeFileTests.cpp [hunk 3]
```diff
@@ -752,3 +752,80 @@ TEST_CASE("MergeFileZip", "[mergefiletest5]") {
...
 }
+
+TEST_CASE("MergeFileOnAttribute", "[mergefiletest5]") {
+  {
+    std::ofstream expectfileFirst(EXPECT_MERGE_CONTENT_FIRST, std::ios::binary);
+    std::ofstream expectfileSecond(EXPECT_MERGE_CONTENT_SECOND, std::ios::binary);
+
+    // Create and write to the test file
+    for (int i = 0; i < 6; i++) {
+      std::ofstream{std::string(FLOW_FILE) + "." + std::to_string(i) + ".txt", std::ios::binary} << std::to_string(i);
```
Comment (hunyadi-dev) [inline libminifi/test/archive-tests/MergeFileTests.cpp:763]: Is there any advantage of performing this in binary mode?

File: libminifi/test/archive-tests/MergeFileTests.cpp [hunk 4]
```diff
@@ -752,3 +752,80 @@ TEST_CASE("MergeFileZip", "[mergefiletest5]") {
...
+  }
+  // validate the merge content
+  std::set<std::shared_ptr<core::FlowFile>> expiredFlowRecords;
+  std::shared_ptr<core::FlowFile> flow1 = output->poll(expiredFlowRecords);
+  std::shared_ptr<core::FlowFile> flow2 = output->poll(expiredFlowRecords);
+  {
+    FixedBuffer callback(flow1->getSize());
+    sessionGenFlowFile.read(flow1, &callback);
+    std::ifstream file1(EXPECT_MERGE_CONTENT_FIRST, std::ios::binary);
+    std::string contents((std::istreambuf_iterator<char>(file1)), std::istreambuf_iterator<char>());
```
Comment (hunyadi-dev) [inline libminifi/test/archive-tests/MergeFileTests.cpp:820]: Again, any advantage of reading `char`-s in binary mode?

File: libminifi/test/archive-tests/MergeFileTests.cpp (hunk 3 above)
Comment (adamdebreceni) [reply libminifi/test/archive-tests/MergeFileTests.cpp:763]: we should always read/write files in binary mode, unless we are up for some surprises, e.g. in non-binary mode on windows every '\n' character is prepended with a '\r'

Comment (hunyadi-dev) [reply libminifi/test/archive-tests/MergeFileTests.cpp:763]: Well, I think this tests should not hide those differences. But we do not have whitespaces here at all, so the difference should not be relevant.

File: libminifi/test/archive-tests/MergeFileTests.cpp [hunk 5]
```diff
@@ -752,3 +752,80 @@ TEST_CASE("MergeFileZip", "[mergefiletest5]") {
...
+
+  core::ProcessSession sessionGenFlowFile(context);
+  std::shared_ptr<core::FlowFile> record[6];
+
+  // Generate 6 flowfiles, even files are merged to one, odd files are merged to an other
+  for (int i = 0; i < 6; i++) {
+    std::shared_ptr<core::FlowFile> flow = std::static_pointer_cast < core::FlowFile > (sessionGenFlowFile.create());
+    std::string flowFileName = std::string(FLOW_FILE) + "." + std::to_string(i) + ".txt";
+    sessionGenFlowFile.import(flowFileName, flow, true, 0);
+    if (i % 2 == 0)
```
Comment (hunyadi-dev) [inline libminifi/test/archive-tests/MergeFileTests.cpp:791]: Minor, but would be more readable as:
```c++
flow->setAttribute("tag", i % 2 ? "odd" : "even");
```

File: libminifi/test/archive-tests/MergeFileTests.cpp [hunk 6]
```diff
@@ -752,3 +752,80 @@ TEST_CASE("MergeFileZip", "[mergefiletest5]") {
...
+  context->setProperty(org::apache::nifi::minifi::processors::MergeContent::MergeStrategy, MERGE_STRATEGY_BIN_PACK);
+  context->setProperty(org::apache::nifi::minifi::processors::MergeContent::DelimiterStrategy, DELIMITER_STRATEGY_TEXT);
+  context->setProperty(org::apache::nifi::minifi::processors::MergeContent::MinEntries, "3");
+  context->setProperty(org::apache::nifi::minifi::processors::MergeContent::CorrelationAttributeName, "tag");
+
+  core::ProcessSession sessionGenFlowFile(context);
+  std::shared_ptr<core::FlowFile> record[6];
+
+  // Generate 6 flowfiles, even files are merged to one, odd files are merged to an other
+  for (int i = 0; i < 6; i++) {
```
Comment (hunyadi-dev) [inline libminifi/test/archive-tests/MergeFileTests.cpp:787]: What is the point of having this separated out from the one above? The first one has no dependency on any state and is directly used here.

File: libminifi/test/archive-tests/MergeFileTests.cpp (hunk 1 above)
Comment (fgerlits) [reply libminifi/test/archive-tests/MergeFileTests.cpp:819]: ok

**Comment by adamdebreceni** [comment]
the shortcomings of this test and other MergeFileTests and CompressContentTests will be addressed in [this](https://issues.apache.org/jira/browse/MINIFICPP-1308) ticket

**Pull request closed without merging**